uploadservice-s3/src/main/java/net/gotev/uploadservice/s3/S3UploadTask.kt

```diff
@@ -0,0 +1,97 @@
+package net.gotev.uploadservice.s3
+
+import com.amazonaws.mobileconnectors.s3.transferutility.TransferState
+import com.amazonaws.regions.Regions
+import net.gotev.uploadservice.UploadTask
+import net.gotev.uploadservice.data.UploadFile
+import net.gotev.uploadservice.logger.UploadServiceLogger
+import net.gotev.uploadservice.network.HttpStack
+import net.gotev.uploadservice.network.ServerResponse
+
+class S3UploadTask() : UploadTask(), S3ClientWrapper.Observer {
+
+    private var uploadBytes: Long = 0
+
+    private val s3params by lazy {
+        S3UploadTaskParameters.createFromPersistableData(params.additionalParameters)
+    }
+    @Throws(Exception::class)
+    override fun upload(httpStack: HttpStack) {
+        val s3params = s3params
+        S3ClientWrapper(uploadId = params.id,
+                context = context,
+                identityPoolId = s3params.identityPoolId,
+                region = Regions.fromName(s3params.region),
+                observer = this
+        ).use { s3Client ->
+
+
+            // this is needed to calculate the total bytes and the uploaded bytes, because if the
+            // request fails, the upload method will be called again
+            // (until max retries is reached) to retry the upload, so it's necessary to
+            // know at which status we left, to be able to properly notify further progress.
+            calculateUploadedAndTotalBytes();
+
+
+            for (file in params.files) {
+                if (!shouldContinue)
+                    break
+
+                if (file.successfullyUploaded)
+                    continue
+
+                s3Client.uploadFile(context, s3params.bucketName, s3params.serverSubpath, file)
+            }
+        }
+    }
+
+    /**
+     * Calculates the total bytes of this upload task.
+     * This the sum of all the lengths of the successfully uploaded files and also the pending
+     * ones.
+     */
+    private fun calculateUploadedAndTotalBytes() {
+        resetUploadedBytes()
```

Comment:
what is this, where is it defined and what does it do? is it required? this seems like it will cause the UI to jump around. Why is this necessary? It seems like all you need to do is calculate the total bytes and call onProgress

Reply:
This is not my code. I don't know what happens if we remove it. I just kept it the same as FTP.

https://github.com/gotev/android-upload-service/blob/ff518f03159c5b1e7b2b001c75d26fbcc22c542a/uploadservice-ftp/src/main/java/net/gotev/uploadservice/ftp/FTPUploadTask.kt#L80

Reply:
But generally, the uploaded bytes is cumulative: uploadedBytes += bytesSent
if we don't reset it each time and if it has some values previously it makes it buggy. This method basically resets uploaded bytes and sums again how many bytes are sent. There might be a better implementation but that probably requires changing the base UploadTask class.